lib/generator/binary_protocol.ex

```diff
@@ -1,57 +1,102 @@
+  You could visual this as a tree of types like the following.
+
+    struct (MyStruct)
+    ├ i32
+    ├ list
+    │ └ i32
+    └ map
+      ├ key
+      │ └ string
+      └ value
+        └ struct (OtherStruct)
+
+  We care about the edges of this graph. This module needs to know how to
+  implement deserializers for each transition from one type to another.
+
+  - struct -> i32
+  - struct -> list
+  - list -> i32
+  - struct -> map
+  - map key -> string
+  - map value -> struct
+
+  For the struct at the root of the graph, we generate a deserializer for each
+  field. Other structs are leaf nodes in the graph. Rather than generating the
+  deserialization logic inline, we make a call to the module we expect to have
+  been generated for that struct.
+  """
+
   alias Thrift.Parser.FileGroup
   alias Thrift.Parser.Models.{
     # Exception,
-    # Field,
+    Field,
     Struct,
     StructRef,
     TEnum,
   }
 
-  def struct_deserializer(%{fields: fields}, name, file_group) do
-    # field_matchers = Enum.map(fields, fn %Field{name: name} ->
-    #   {name, Macro.var(name, nil)}
-    # end)
-
-    # struct_matcher = {:%, [], [name, {:%{}, [], field_matchers}]}
+  @doc """
+  Generate a deserializer for a Thrift struct or exception.
 
-    # field_serializers = Enum.map(fields, fn %Field{name: name, type: type, id: id} ->
-    #   var = Macro.var(name, nil)
-    #   quote do
-    #     case unquote(var) do
-    #       nil ->
-    #         []
-    #       _ ->
-    #         unquote([
-    #           quote do <<unquote(type_id(type, file_group)), unquote(id) :: size(16)>> end,
-    #           value_serializer(type, var, file_group)
-    #         ] |> merge_binaries)
-    #     end
-    #   end
-    # end)
+  At the moment it also generates an experimental serializer that may be faster.
+  """
+  def struct_deserializer(%{fields: fields}, name, file_group) do
+    field_matchers = Enum.map(fields, fn %Field{name: name} ->
+      {name, Macro.var(name, nil)}
+    end)
+
+    struct_matcher = {:%, [], [name, {:%{}, [], field_matchers}]}
+
+    field_serializers = Enum.map(fields, fn %Field{name: name, type: type, id: id} ->
+      var = Macro.var(name, nil)
+      quote do
+        case unquote(var) do
+          nil ->
+            []
+          _ ->
+            unquote([
+              quote do <<unquote(type_id(type, file_group)), unquote(id) :: size(16)>> end,
+              value_serializer(type, var, file_group)
+            ] |> merge_binaries)
+        end
+      end
+    end)
 
     field_deserializers = fields
     |> Enum.map(&field_deserializer(&1.type, &1, :deserialize, file_group))
     |> merge_blocks
 
     quote do
-      # def serialize(unquote(struct_matcher)) do
-      #   unquote([field_serializers, <<0>>] |> merge_binaries |> merge_binaries)
-      # end
+      # defp bool_to_int(nil), do: 0
+      # defp bool_to_int(false), do: 0
+      # defp bool_to_int(_), do: 1
+
+      def serialize2(unquote(struct_matcher)) do
```

Comment:
I uncommented this stuff to do some more benchmarking. If it's faster, we ought to use it.